### PR TITLE
Inline types not exported from their erlang modules

### PR DESCRIPTION
**Pull request by michallepicki**
This commit fixes the following errors from dialyzer analysis:
```
Unknown types:
  beam_lib:chnk_rsn/0
  erlang:message_queue_data/0
  supervisor:modules/0
  supervisor:restart/0
  supervisor:shutdown/0
```

I'm not sure if this is the right way to fix these. In my view, using unexported types is similar to accessing private APIs. Maybe there is no reason for these types not to be exported and this should be fixed upstream by exporting them? Or maybe this is a hint that Elixir is indeed accessing "private" APIs in some of these cases? Or maybe this approach is fine?

### CODE REVIEW

File: lib/mix/lib/mix/release.ex [hunk 1]
```diff
@@ -777,7 +777,21 @@ defmodule Mix.Release do
...
+          | {:invalid_chunk, :file.filename(), chunkid()}
+          | {:missing_chunk, :file.filename(), chunkid()}
+          | {:not_a_beam_file, :file.filename()}
+          | {:file_error, :file.filename(), :file.posix()}
+  @type chnk_rsn() ::
+          {:unknown_chunk, :file.filename(), atom()}
+          | {:key_missing_or_invalid, :file.filename(), :abstract_code | :debug_info}
+          | info_rsn()
+  # inlining chnk_rsn() here because :beam_lib.chnk_rsn() is not exported
+  @spec strip_beam(binary()) :: {:ok, binary} | {:error, :beam_lib, chnk_rsn()}
```
Comment (josevalim) [inline lib/mix/lib/mix/release.ex:794]:
```suggestion
  @spec strip_beam(binary()) :: {:ok, binary} | {:error, :beam_lib, term}
```

**Comment by josevalim** [comment]
I think those are good, thanks! But the chunk_rsn is leaking a bunch of beam_lib implementation details, so I think it is just better to say `term` for now. But ideally I would say all of these should be available upstream, yeah. It is good to merge IMO. :)

**Comment by michallepicki** [comment]
@josevalim I'll remove my comments

**Comment by josevalim** [comment]
@michallepicki I think the comments are actually good. They are inform there may be a better solution upstream, so I would keep them. :)

**Comment by michallepicki** [comment]
@josevalim do we have a system for "tags" inside comments? Like an easily searchable identifier `# %inlined-erlang-type%` and a longer explaination in one of the places?

**Comment by josevalim** [comment]
We have TODO, although I don't think they apply here.

**Comment by josevalim** [comment]
To be honest, I think the comments were totally fine as is. :)

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:

**Pull request merged**